includes/express-checkout/class-wc-payments-express-checkout-button-helper.php

```diff
@@ -760,22 +760,33 @@ private function is_product_supported() {
 		 *
 		 * @psalm-suppress UndefinedClass
 		 */
-		if ( is_null( $product )
-			|| ! is_object( $product )
-			|| ! in_array( $product->get_type(), $this->supported_product_types(), true )
-			|| ( class_exists( 'WC_Subscriptions_Product' ) && $product->needs_shipping() && WC_Subscriptions_Product::get_trial_length( $product ) > 0 ) // Trial subscriptions with shipping are not supported.
+
+		if ( is_null( $product ) || ! is_object( $product ) ) {
+			$is_supported = false;
+		} else {
+			// Simple subscription that needs shipping with free trials is not supported.
+			$is_free_trial_simple_subs = class_exists( 'WC_Subscriptions_Product' ) && $product->get_type() === 'subscription' && $product->needs_shipping() && WC_Subscriptions_Product::get_trial_length( $product ) > 0;
+
+			// Disable ECE for all variable subscriptions with free trials, as they are not currently supported. For now, ECE will be disabled for all such cases, and a solution will be addressed in a separate PR.
+			$is_free_trial_variable_sub = class_exists( 'WC_Subscriptions_Product' ) && $product->get_type() === 'variable-subscription' && WC_Subscriptions_Product::get_trial_length( $product ) > 0;
+
+			if (
+			! in_array( $product->get_type(), $this->supported_product_types(), true )
+			|| $is_free_trial_simple_subs
+			|| $is_free_trial_variable_sub
```

Comment:
These are the important lines to review; the other changes were just to try to simplify the code a bit. Reviewing with whitespace changes hidden may be helpful.